### PR TITLE
feat(platform): unified cross-platform abstraction layer (Phases 0–5)

### PR DESCRIPTION
**Pull request by mweinbach**
## Why

This repo suffered chronic **fix-macOS-break-Windows ping-pong**: the same platform-sensitive concern was re-implemented N times (5 executable resolvers, 5 home-dir precedences, 4 graceful-kill escalations, 3 atomic-write strategies, 14 hand-synced copies of Windows shell prose), so a fix landed in one copy while the others kept the old behavior. A 40-agent audit of `main` mapped **253 verified cross-platform hazards across 83 divergence clusters**; this PR builds the unified platform layer that ends it and migrates the highest-value call sites onto it.

The rule the layer enforces: **one implementation per platform-sensitive concern, living under `src/platform/`. Platform branching happens inside those modules; callers never read `process.platform`.** A ratchet test bans new violations and only counts down.

Full audit + migration map: [`docs/platform-abstraction-plan.md`](docs/platform-abstraction-plan.md).

## What changed (Phases 0–5)

**The layer** — `src/platform/{host,pathString,text,paths,env,exec,proc,fs,shell,approval,sandbox,pathImpl}.ts` + barrel. Every platform-branching function takes `platform?: NodeJS.Platform` so all branches are unit-testable on every host.

**Guardrail** — `.gitattributes` pins repo bytes to LF, and `test/platform-boundary.test.ts` bans `process.platform`/`os.homedir`/`os.tmpdir`/`Bun.which`/`process.arch`/`path.win32|posix` outside `src/platform/**` (and the layer's own tests). A checked-in baseline grandfathers current offenders and only shrinks — it already dropped as migrations removed tokens.

**The worst confirmed bugs, fixed and verified:**

- **CRITICAL — multi-line edits failed on every CRLF checkout.** `read` strips `\r`, `edit` matched raw bytes, so any multi-line `oldString` the model copied from `read` output could never match on Windows (`core.autocrlf=true`), and successful edits spliced LF into CRLF files. `edit` now matches LF-normalized and re-emits the file's dominant EOL; `read` gained BOM/UTF-16 decoding.
- **Windows quoting corruption** — the bash tool sends commands via `pwsh -EncodedCommand` (one parse layer on every OS, `-File` fallback above the CreateProcess ceiling, pinned exit-code contract). Live-verified against real pwsh: `git commit -m` with nested quotes round-trips byte-exact, `cmd /c exit 3` propagates as 3 (was 1), Unicode survives.
- **Missing Windows approval gate** — `Remove-Item -Recurse -Force`, `rd /s /q`, `rm -r -fo`, `Format-Volume` etc. now prompt on Windows, the dialect its sessions are steered into; the POSIX destructive set stays a shared floor everywhere.
- **Sandboxed commands losing credentials on Windows** — the sandbox child-env allowlist now includes `USERPROFILE`/`APPDATA`/`LOCALAPPDATA`/`ProgramFiles`/`PYTHONUTF8`; both bash lanes resolve the shell through the same `exec.which`.
- **Bash timeouts orphaned grandchildren** — `proc.run` spawns POSIX children detached and tree-kills on timeout/abort (win32 `taskkill /T`); proven by a live grandchild-heartbeat test.
- **Security case-folding** — the `.GIT`/`.COWORK` protected-metadata bypass is closed on case-insensitive filesystems, and credential-deny canonicalization is unified on the native-realpath engine so a cased `.COWORK/AUTH` can't evade the deny on macOS/Windows.
- **Two Windows HIGH one-liners** — conversation import uses `platform.home()` (was a silent no-op on Windows via `env.HOME ?? cwd`); MCP stdio children merge onto the platform-safe env instead of replacing it (kept `SystemRoot`/`PATH`/`APPDATA`).

**Model-as-compat-layer removed** — prompts now carry only the *host's* shell dialect rendered from `shell.promptGuidance` (14 prose copies → 1 function); `py -3` is deleted everywhere in favor of `shell.pythonInvocation` (the registry-resolved launcher bypassed the runtime PATH prelude, the Windows-only missing-modules class); Codex turns strip the shell section entirely.

**Also:** sandbox scratch-root parity across all three backends + memoized Windows probe + win32/WSA denial markers; glob/grep/ripgrep pattern normalization preserving POSIX escapes; `fs.ts` atomic write/replace/move/lock primitives; portable `execFileCompat` fixtures.

## Verification

Full CI lane on Windows (`bun test --max-concurrency 1`, `bun run typecheck`, `bun run lint`, `bun run docs:check`): typecheck/lint/docs clean. Test failures went from the **24-item pre-change baseline to 19 — zero new regressions**, and 5 `execFileCompat` Windows failures are now fixed. The 19 remaining are all pre-existing baseline failures untouched by this work (lmstudio path separators, markdown-media, workflow YAML, conversation-import Windows decoding — all later-phase migration targets). Every new module ships exhaustive platform-parameterized tests; the whole set was adversarially reviewed (verdict: ship).

## Reviewer notes / not in this PR

- **Deferred (needs live Electron verification):** the `server/shutdown` JSON-RPC route + desktop graceful-kill migration (plan rows 26-27). The `proc.terminateGracefully` primitive it consumes is already here.
- **Pre-merge gate:** the native Windows-sandbox enforcement integration lane (`RUN_WINDOWS_SANDBOX_INTEGRATION=1`) requires a one-time UAC/WFP install and can't run headless — it should be run on a native Windows host before merge, since this PR changes the `no-project-write` helper mode-mapping.
- **Fast-follow items** (non-blocking, tracked): a stale-lock steal-race window in `fs.acquireLockDir`, and `windowsVerbatimArguments` not carried through the legacy `execFileCompat` batch-shim path (self-resolves when the remaining call sites move to `proc.run`).
- One pre-existing latent test race (`research.service.followup`) was hardened to wait for its side-effect — identical structure exists on `main`; this PR's added test files shifted suite timing enough to expose it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches security-sensitive paths (sandbox env, approval classification, protected-metadata/containment semantics), core command execution and process lifecycle, and broad cross-platform behavior where regressions are hard to see without full OS matrix testing.
> 
> **Overview**
> This PR adds a **unified platform layer** under `src/platform/` (host, paths, text, env, exec, proc, fs, shell, approval, sandbox) plus a **barrel export** and **`docs/platform-abstraction-plan.md`**, with **Phase 0** repo hygiene via root **`.gitattributes`** (LF text, CRLF for `.bat`/`.cmd`, binary exceptions) and a **platform-boundary ratchet test** that blocks new raw `process.platform` / path / env usage outside the layer.
> 
> **Agent-facing and runtime behavior** is rewired onto that layer: the **bash tool** uses shared shell transport (including Windows **`-EncodedCommand`**), **`exec.which`** on both sandboxed and unsandboxed lanes, expanded **sandbox child env** on Windows, **`proc.run` tree-kill** on timeout, and **per-dialect destructive-command approval** (PowerShell verbs no longer auto-approve). **`edit`/`read`** share an EOL contract via **`text.replaceRespectingEol`** and BOM/UTF-16 decoding; **MCP stdio** spawns with **`childEnv`** merges instead of replacing the process environment.
> 
> **Prompts and harness** stop duplicating Windows shell/Python rules: model templates drop hand-written PowerShell/`py -3` lines in favor of centralized **`shell.promptGuidance`** / **`pythonInvocation`**, and **`packages/harness`** re-exports canned commands from **`src/platform/shell`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 18938b1a34f433351da0afbe7c359afdd838daab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/tools/edit.ts [hunk 1]
```diff
@@ -66,21 +76,28 @@ export function createEditTool(ctx: ToolContext) {
...
+        // Defensive: the pre-checks above mirror replaceRespectingEol's rules.
+        throw new Error(
+          result.reason === "not_found"
+            ? `oldString not found in ${abs}`
+            : `oldString is not unique in ${abs}. Provide more context or set replaceAll=true.`,
+        );
+      }
       await ctx.assertCanMutate?.("edit");
-      await Bun.write(abs, content);
+      await Bun.write(abs, result.content);
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/tools/edit.ts]: Fixed in 03fdff20. Edits now preserve the decoded file's UTF-16LE/UTF-16BE encoding and BOM instead of rewriting as UTF-8. Regression tests cover both byte orders and CRLF content.

File: src/tools/grep.ts [hunk 2]
```diff
@@ -107,7 +111,32 @@ export function createGrepTool(
...
+      // would fail or silently mangle arguments. Route it through the shim-aware
+      // spawn planner, which either builds a safe cmd.exe wrapper or throws a
+      // typed UnsafeShimArgumentError for arguments cmd.exe cannot carry safely.
+      let spawnFile = rgPath;
+      let spawnArgs = args;
+      if (classifyExecutable(rgPath) === "batch-shim") {
+        try {
+          const plan = resolveSpawn(rgPath, args);
+          spawnFile = plan.file;
+          spawnArgs = plan.args;
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/tools/grep.ts]: Fixed in 03fdff20. `execFileCompat` now carries `windowsVerbatimArguments`, and grep forwards the resolved spawn plan flag. The simulated-win32 regression asserts the cmd wrapper/verbatim contract and the unsafe-shim rejection path.

File: src/tools/edit.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/tools/edit.ts]: Fixed in 03fdff20. The edit tool now reads raw bytes, decodes through the shared UTF-8/UTF-16 contract, and re-encodes with the original UTF-16LE/UTF-16BE BOM and newline form preserved. Regression coverage exercises both UTF-16 byte orders.

**Review by mweinbach — COMMENTED** [review]
Reviewed the updated platform abstraction head and pushed the fixes in `03fdff20`.

Beyond the three existing threads, the pass fixed five additional cross-platform issues: retained POSIX process-group hard-kill escalation, rejected Windows batch-shim shell candidates, streamed UTF-16 reads with bounded early termination, honored injected POSIX `HOME`, and safely quoted managed Python paths containing PowerShell metacharacters.

Verification: targeted regressions pass (338 pass / 5 skip, plus process-tree tests 3 pass / 3 skip); `bun run check`, `bun run typecheck`, `bun run lint`, and `bun run docs:check` pass. Exact-head CI is running.

File: src/platform/proc.ts [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+      void killTreeWin32(proc.pid);
+      return;
+    }
+    killPosixGroup(proc.pid, killSignal, defaultKill);
+    if (killSignal !== "SIGKILL") {
+      escalationTimer = setTimeout(
+        () => killPosixGroup(proc.pid, "SIGKILL", defaultKill),
+        POSIX_HARD_KILL_ESCALATION_MS,
+      );
+      escalationTimer.unref?.();
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/platform/proc.ts:318]: Valid finding; fixed in c1cd54a0. The hard escalation now targets only the detached POSIX process group and never falls back to the original root PID. Root exit immediately disarms the delayed timer and reaps surviving descendants, so no stale PID timer survives `run()`. The regression asserts an ESRCH group lookup produces no positive-PID kill, while the TERM-ignoring descendant test still passes.

File: src/platform/fs.ts [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+    const record: LockDirOwner = {
+      pid: process.pid,
+      hostname: os.hostname(),
+      acquiredAt,
+      heartbeatAt: new Date(now()).toISOString(),
+    };
+    await writeFileAtomic(ownerPath, JSON.stringify(record), {}, deps);
+  };
+  try {
+    await writeOwner();
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/platform/fs.ts]: Valid finding; fixed in aedd83ac. Initial `owner.json` creation is now exclusive and carries a per-acquisition token; the acquirer reads the token back before returning a handle. If an ownerless directory is replaced mid-write, the delayed writer detects `LOCK_OWNERSHIP_LOST` and cannot overwrite or remove the replacement owner's lock. A deterministic gated-write regression covers the race.

File: src/platform/exec.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  } = {},
+): SpawnPlan {
+  const platform = opts.platform ?? hostPlatform();
+  if (platform !== "win32") {
+    return { file: fileOrName, args: [...args], kind: "native" };
+  }
+  const resolved = which(fileOrName, { ...opts, platform }) ?? fileOrName;
+  const kind = classifyExecutable(resolved, platform);
+  if (kind !== "batch-shim") {
+    return { file: fileOrName, args: [...args], kind };
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/platform/exec.ts]: Valid finding; fixed in 12685969. Win32 `resolveSpawn()` now returns the path found by `which()` for native executables and PowerShell scripts, while preserving the original name only when lookup fails. Unit coverage now asserts the resolved `.exe` path, including the `skipDirs` fallback case.

File: src/tools/grep.ts [hunk 6]
```diff
@@ -107,10 +117,38 @@ export function createGrepTool(
...
+          }
+          throw err;
+        }
+      }
+
+      const result = await execFileImpl(spawnFile, spawnArgs, {
         maxBuffer: 1024 * 1024 * 10,
         ...(ctx.abortSignal ? { signal: ctx.abortSignal } : {}),
         timeoutMs,
+        windowsVerbatimArguments,
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/tools/grep.ts:154]: Valid finding; fixed in 18938b1a. Production grep execution now uses `platform.proc.run` on every host, so timeout/abort tree-kills descendants, including the `cmd.exe -> rg.exe` batch-shim case on Windows. `execFileImpl` remains only as the deterministic unit-test seam, and the simulated-win32 test asserts the tree-aware runner receives the shim plan and platform.

File: src/platform/shell.ts [hunk 7]
```diff
@@ -34,43 +77,206 @@ function envValue(env: Record<string, string | undefined>, name: string): string
...
+    };
+  }
+  return {
+    runPythonScript: (scriptPath) => `${q(python.command)} ${q(scriptPath)}`,
+    printWorkingDirectory: () => "pwd",
+    listDirectory: (dirPath) => (dirPath ? `ls -la ${q(dirPath)}` : "ls -la"),
+    countLines: (filePath) => `wc -l ${q(filePath)}`,
+  };
+}
+
```
[the remark this answers is not in the document]
Comment (mweinbach) [reply src/platform/shell.ts:266]: Valid finding; fixed in b75c3b95. `buildPlatformShellCommandWithRuntimePrelude()` now delegates runtime directory assembly and platform-aware deduplication to the canonical `env.runtimePathDirs()` implementation. Existing Windows/POSIX PATH-prelude and runtime-env suites pass.

**Review by mweinbach — COMMENTED** [review]
Exact-head review complete for `18938b1a34f433351da0afbe7c359afdd838daab`.

All 8 inline review threads were validated, fixed where applicable, replied to with commit/test evidence, and resolved. The fixes cover UTF-16-safe editing and streaming reads, Windows batch-shim/PowerShell spawning, native executable resolution, POSIX process-tree escalation, lock ownership races, canonical runtime PATH assembly, and grep timeout/abort tree cleanup. The review also caught and fixed the exact-head CI regressions plus independent cross-platform edge cases.

Verification on this exact SHA is green: unit tests, lint/typecheck/docs, build smoke, remote MCP smoke, mobile, macOS ARM64, Windows x64, and Windows ARM64. Cursor Bugbot completed successfully; the final heartbeat found zero unresolved threads and zero new external comments. The PR is merge-clean.
